File: nanopay/src/net/nanopay/invoice/InvoiceNotificationDAO.java
package net.nanopay.invoice;

import foam.core.FObject;
import foam.core.X;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.nanos.app.AppConfig;
import foam.nanos.auth.User;
import foam.nanos.auth.token.TokenService;
import foam.nanos.notification.email.EmailMessage;
import foam.nanos.notification.email.EmailService;
import foam.util.SafetyUtil;
import java.text.SimpleDateFormat;
import java.util.*;
import net.nanopay.auth.PublicUserInfo;
import net.nanopay.accounting.quickbooks.model.QuickbooksInvoice;
import net.nanopay.accounting.xero.model.XeroInvoice;
import net.nanopay.invoice.model.Invoice;
import net.nanopay.invoice.model.InvoiceStatus;
import net.nanopay.invoice.notification.NewInvoiceNotification;
import net.nanopay.model.Currency;

/**
 * Invoice decorator for dictating and setting up new invoice notifications and emails.
 * Responsible for sending notifications to both internal and external users on invoice create.
 */
public class InvoiceNotificationDAO extends ProxyDAO {

  protected DAO bareUserDAO_;
  protected DAO notificationDAO_;
  protected DAO currencyDAO_;
  protected TokenService externalToken;

  public InvoiceNotificationDAO(X x, DAO delegate) {
    super(x, delegate);
    bareUserDAO_ = (DAO) x.get("bareUserDAO");
    notificationDAO_ = (DAO) x.get("notificationDAO");
    currencyDAO_ = (DAO) x.get("currencyDAO");
    externalToken = (TokenService) x.get("externalInvoiceToken");
  }

  @Override
  public FObject put_(X x, FObject obj) {
    Invoice invoice = (Invoice) obj;
    Invoice existing = (Invoice) super.find(invoice.getId());

    // if the invoice is imported from accounting software
    if ( existing == null &&
         (invoice instanceof QuickbooksInvoice || invoice instanceof XeroInvoice) ) {
      return super.put_(x, invoice);
    }

    if ( existing != null ) {
      if (
        existing.getStatus() == InvoiceStatus.UNPAID ||
        existing.getStatus() == InvoiceStatus.PENDING_APPROVAL
      ) {
        if ( invoice.getStatus() == InvoiceStatus.PENDING ) {
          AppConfig appConfig = (AppConfig) x.get("appConfig");
          User user = invoice.findPayeeId(x);
          String emailTemplate = "invoice-paid";

          EmailService emailService = (EmailService) x.get("email");
          EmailMessage message = new EmailMessage.Builder(x)
            .setTo(new String[]{user.getEmail()})
            .build();

          PublicUserInfo payer = invoice.getPayer();

          String amount = ((Currency) currencyDAO_.find(invoice.getDestinationCurrency()))
            .format(invoice.getAmount());

          String currency = invoice.getDestinationCurrency();

          HashMap<String, Object> args = new HashMap<>();
          args.put("amount", amount);
          args.put("currency", currency);
          args.put("name", invoice.getPayee().label());
          args.put("link", appConfig.getUrl());
          args.put("fromName", payer.label());
          args.put("senderCompany", payer.label());

          emailService.sendEmailFromTemplate(x, user, message, emailTemplate, args);
        }
      }
    }

    // Only send invoice notification if invoice does not have a status of draft
    if ( ! InvoiceStatus.DRAFT.equals(invoice.getStatus()) ) {
      // if no existing invoice has been sent OR the existing invoice was a draft, send an email
      if ( existing == null || InvoiceStatus.DRAFT.equals(existing.getStatus()) ) {
        sendInvoiceNotification(x, invoice);
      }
    }

    return super.put_(x, invoice);
  }

  private void sendInvoiceNotification(X x, Invoice invoice) {
    long payeeId = invoice.getPayeeId();
    long payerId = invoice.getPayerId();

    NewInvoiceNotification notification = new NewInvoiceNotification();

    /*
     * If invoice is external, calls the external token service and avoids internal notifications,
     * otherwise sets email args for internal user email and creates notification.
     */
    if ( invoice.getExternal() ) {
      // Sets up required token parameters.
      long externalUserId = invoice.getContactId();
      User externalUser = (User) bareUserDAO_.find(externalUserId);
      Map tokenParams = new HashMap();
      tokenParams.put("invoice", invoice);

      externalToken.generateTokenWithParameters(x, externalUser, tokenParams);
      return;
    } else {
      User user = (User) x.get("user");

      /*
       * For original nanopay app, if current user is equal to payer, it will load
       * the 'payable' email template with `"group":"*"`.
       * For SME/Ablii, if current user is equal to payer, it will load the 'receivable'
       * email template with `"group":"sme"`.
       */
      String template = user.getId() == payerId ? "transfer-paid" : "receivable";

      // Set email values on notification.
      notification = setEmailArgs(x, invoice, notification);
      notification.setEmailName(template);
      notification.setEmailIsEnabled(user.getId() == payerId ? false : true);
    }

    notification.setUserId(payeeId == ((Long)invoice.getCreatedBy()) ? payerId : payeeId);
    notification.setInvoiceId(invoice.getId());
    notification.setNotificationType("Invoice received");
    notificationDAO_.put(notification);
  }

  private NewInvoiceNotification setEmailArgs(X x, Invoice invoice, NewInvoiceNotification notification) {
    SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MMM-YYYY");

    PublicUserInfo payee = invoice.getPayee();
    PublicUserInfo payer = invoice.getPayer();

    // If invType is true, then payee sends payer the email and notification.
    boolean invType = invoice.getPayeeId() == invoice.getCreatedBy();

    // Add the currency symbol and currency (CAD/USD, or other valid currency)
    String amount = ((Currency) currencyDAO_.find(invoice.getDestinationCurrency()))
        .format(invoice.getAmount()) + " " + invoice.getDestinationCurrency();

    String accountVar = SafetyUtil.isEmpty(invoice.getInvoiceNumber()) ? "N/A" : invoice.getInvoiceNumber();

    notification.getEmailArgs().put("amount", amount);
    notification.getEmailArgs().put("account", accountVar);
    notification.getEmailArgs().put("name", invType ? payer.getFirstName() : payee.getFirstName());
    notification.getEmailArgs().put("fromEmail", invType ? payee.getEmail() : payer.getEmail());
    notification.getEmailArgs().put("fromName", invType ? payee.label() : payer.label());

    if ( invoice.getDueDate() != null ) {
      notification.getEmailArgs().put("date", dateFormat.format(invoice.getDueDate()));
    }

    AppConfig appConfig = (AppConfig) x.get("appConfig");
    notification.getEmailArgs().put("link", appConfig.getUrl());
    notification.getEmailArgs().put("senderCompany", invType ? payee.getOrganization() : payer.getOrganization() );
    return notification;
  }
}

File: nanopay/src/net/nanopay/invoice/InvoiceNotificationDAO.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.invoice',
  name: 'InvoiceNotificationDAO',

  extends: 'foam.dao.ProxyDAO',

  javaImports: [
    'foam.core.Currency',
    'foam.core.X',
    'foam.dao.ArraySink',
    'foam.dao.DAO',
    'foam.nanos.app.AppConfig',
    'foam.nanos.auth.Group',
    'foam.nanos.auth.Subject',
    'foam.nanos.auth.User',
    'foam.nanos.auth.UserUserJunction',
    'foam.nanos.auth.token.TokenService',
    'foam.nanos.logger.Logger',
    'foam.nanos.notification.Notification',
    'foam.nanos.notification.email.EmailMessage',

    'java.text.SimpleDateFormat',
    'java.util.*',

    'net.nanopay.invoice.model.BillingInvoice',
    'net.nanopay.invoice.model.Invoice',
    'net.nanopay.invoice.model.InvoiceStatus',
    'net.nanopay.invoice.model.PaymentStatus',
    'net.nanopay.model.Business',
    'static foam.mlang.MLang.*'
  ],

  documentation: `
    Invoice decorator for dictating and setting up new invoice notifications and emails.
    Responsible for sending notifications to both internal and external users on invoice create.
    Triggers on invoices that send emails are as follows:
    1) invoiceIsBeingPaidButNotComplete
    2) invoiceIsANewRecievable
    3) invoiceNeedsApproval
    4) invoiceIsBeingPaidAndCompleted
  `,

  methods: [
    {
      name: 'put_',
      javaCode: `
        // Gathering Variables and checking null objects
        if ( obj == null || ((Invoice) obj).getStatus() == null ) return obj;

        Invoice invoice = (Invoice) obj;
        Invoice oldInvoice = (Invoice) super.find(invoice.getId());
        Subject subject = ((Subject) x.get("subject"));
        User agent = subject.getRealUser();
        User user = subject.getUser();

        String agentName = agent != user ? agent.getFirstName() + " " + agent.getLastName() : null;

        // CPF-1322 showed an issue with an invoice not being saved in dao due to error down chain
        // thus confirm invoice put first.
        invoice = (Invoice) super.put_(x, invoice);
        if ( invoice == null ) return invoice;

        User payerUser = (User) invoice.findPayerId(x);
        User payeeUser = (User) invoice.findPayeeId(x);

        boolean isPayerABusiness = (invoice.findPayerId(x) instanceof Business);
        boolean isPayeeABusiness = (invoice.findPayeeId(x) instanceof Business);

        String businessName = payeeUser instanceof Business ? payeeUser.getBusinessName() : payeeUser.getOrganization();

        InvoiceStatus newInvoiceStatus = invoice.getStatus();
        InvoiceStatus oldInvoiceStatus = oldInvoice != null ? oldInvoice.getStatus() : null;

        // Various condition checks
        boolean invoiceHasBeenMarkedComplete =
          ( oldInvoiceStatus == null || oldInvoiceStatus != InvoiceStatus.PAID )
          &&
          invoice.getPaymentMethod() == PaymentStatus.CHEQUE; //PaymentStatus.CHEQUE is used when we 'Mark as Complete'

        boolean isARecievable = user.getId() == invoice.getPayeeId();
        boolean invoiceIsBeingPaidButNotComplete =
          ( oldInvoiceStatus == null || oldInvoiceStatus != InvoiceStatus.PROCESSING )
          &&
          ( newInvoiceStatus == InvoiceStatus.PROCESSING )
          &&
          invoice.getPaymentDate() != null
          &&
          invoice.isPropertySet("paymentId");
        boolean invoiceIsBeingPaidAndCompleted =
          ( oldInvoiceStatus == null || oldInvoiceStatus != InvoiceStatus.PAID )
          &&
            newInvoiceStatus == InvoiceStatus.PAID
          &&
          invoice.getPaymentDate() != null
          &&
          invoice.getPaymentMethod() != PaymentStatus.CHEQUE; //PaymentStatus.CHEQUE is used when we 'Mark as Complete'
        boolean invoiceNeedsApproval =
            ( oldInvoiceStatus == null || oldInvoiceStatus != InvoiceStatus.PENDING_APPROVAL )
          &&
            newInvoiceStatus == InvoiceStatus.PENDING_APPROVAL;
        boolean invoiceIsARecievable =
          ( oldInvoiceStatus == null || oldInvoiceStatus != InvoiceStatus.UNPAID )
          &&
            newInvoiceStatus == InvoiceStatus.UNPAID
          &&
          isARecievable;
        boolean invoiceIsPartOfFeesScheduledInvoice =
          oldInvoiceStatus == null
          &&
          newInvoiceStatus == InvoiceStatus.SCHEDULED;

        // Performing Actions based on whats been set to true.
        if ( invoiceIsBeingPaidButNotComplete || invoiceIsARecievable || invoiceNeedsApproval || invoiceIsBeingPaidAndCompleted || invoiceHasBeenMarkedComplete || invoiceIsPartOfFeesScheduledInvoice) {
          String[] emailTemplates = { "payable", "receivable", "invoice-approval-email", "invoice-transaction-completed", "mark-as-complete", "scheduledEmail" };
          HashMap<String, Object> args = null;
          boolean invoiceIsToAnExternalUser = invoice.getExternal();
          DAO currencyDAO = (DAO) x.get("currencyDAO");
          TokenService externalInvoiceToken = (TokenService) x.get("externalInvoiceToken");

          // FIELD LIST FOR:
          // populateArgsForEmail(args(Map), invoice, fromName, dueDate(bool), currencyDAO, agentName, invoiceType)
          // sendEmailFunction(x, isContact, emailTemplateName, invoiceId, userBeingSentEmail, args(Map), externalInvoiceToken)

          try {
            if ( invoiceIsBeingPaidButNotComplete ) {
              args = populateArgsForEmail(args, invoice, payerUser.toSummary(), invoice.getCreated(), currencyDAO, agentName, null);
              sendEmailFunction(x, invoiceIsToAnExternalUser, emailTemplates[0], invoice.getId(), payeeUser, args, externalInvoiceToken );
            }
            if ( invoiceIsARecievable ) {
              args = populateArgsForEmail(args, invoice, payeeUser.toSummary(), invoice.getDueDate(), currencyDAO, agentName, null);
              sendEmailFunction(x, invoiceIsToAnExternalUser, emailTemplates[1], invoice.getId(), payerUser, args, externalInvoiceToken );
            }
            if ( invoiceNeedsApproval ) {
              User tempApprover = null;
              User currentUser = ((Subject) x.get("subject")).getUser();
              List<UserUserJunction> approvers = findApproversOftheBusiness(x);
              DAO userDAO = (DAO) x.get("userDAO");
              for ( UserUserJunction approver : approvers ) {
                if ( args != null ) args.clear();
                tempApprover = (User) userDAO.find(approver.getPartnerId());
                if ( tempApprover == null ) continue;
                args = populateArgsForEmail(args, invoice, agent.getFirstName(), invoice.getDueDate(), currencyDAO, agentName, null);
                args.put("paymentTo", payeeUser.toSummary());
                sendEmailFunction(x, false, emailTemplates[2], invoice.getId(), tempApprover, args, externalInvoiceToken);
              }
            }
            if ( invoiceIsBeingPaidAndCompleted ) {
              args = populateArgsForEmail(args, invoice, payerUser.toSummary(), invoice.getPaymentDate(), currencyDAO, agentName, null);
              sendEmailFunction(x, invoiceIsToAnExternalUser, emailTemplates[3], invoice.getId(), payeeUser, args, externalInvoiceToken );
            }
            if ( invoiceHasBeenMarkedComplete ) {
              args = populateArgsForEmail(args, invoice, agent.getFirstName(), invoice.getPaymentDate(), currencyDAO, agentName, "payable");
              sendEmailFunction(x, invoiceIsToAnExternalUser, emailTemplates[4], invoice.getId(), payerUser, args, externalInvoiceToken );
            }
            if ( invoiceIsPartOfFeesScheduledInvoice ) {
              args = populateArgsForEmail(args, invoice, payeeUser.toSummary(), invoice.getPaymentDate(), currencyDAO, agentName, null);
              sendEmailFunction(x, invoiceIsToAnExternalUser, emailTemplates[5], invoice.getId(), payerUser, args, externalInvoiceToken );
            }
          } catch (Exception e) {}
        }
        return invoice;
      `
    },
    {
      name: 'sendEmailFunction',
      args: [
        {
          name: 'x',
          type: 'Context'
        },
        {
          name: 'isContact',
          class: 'Boolean'
        },
        {
          name: 'emailTemplateName',
          class: 'String'
        },
        {
          name: 'invoiceId',
          class: 'Long'
        },
        {
          name: 'userBeingSentEmail',
          type: 'User'
        },
        {
          name: 'args',
          class: 'Map',
          javaType: 'java.util.HashMap<String, Object>',
        },
        {
          name: 'externalInvoiceToken',
          type: 'TokenService'
        }
      ],
      javaCode: `
        Group group = (Group) userBeingSentEmail.findGroup(x);
        AppConfig appConfig = group.getAppConfig(x);
        args.put("link", appConfig.getUrl());
        args.put("name", User.FIRST_NAME);
        args.put("sendTo", User.EMAIL);
        args.put("invoiceId", invoiceId);

        if ( isContact ) {
          args.put("template", emailTemplateName);
          externalInvoiceToken.generateTokenWithParameters(x, userBeingSentEmail, args);
        } else {
          Notification notification = new Notification.Builder(x)
            .setBody("Invoice Notification.")
            .setNotificationType(emailTemplateName + " email.")
            .setEmailArgs(args)
            .setEmailName(emailTemplateName)
            .build();

          userBeingSentEmail.doNotify(x, notification);
        }
      `
    },
    {
      name: 'populateArgsForEmail',
      javaType: 'java.util.HashMap<String, Object>',
      args: [
        {
          name: 'args',
          class: 'Map',
          javaType: 'java.util.HashMap<String, Object>',
        },
        {
          name: 'invoice',
          type: 'Invoice'
        },
        {
          name: 'fromName',
          class: 'String'
        },
        {
          name: 'date',
          class: 'Date'
        },
        {
          name: 'currencyDAO',
          type: 'DAO'
        },
        {
          class: 'String',
          name: 'agentName'
        },
        {
          class: 'String',
          name: 'invoiceType'
        }
      ],
      javaCode: `
      HashMap<String, Object> newArgs = new HashMap<>();
      newArgs.put("agentName", agentName);
      newArgs.put("fromName", fromName);
      newArgs.put("invoiceType", invoiceType);
      newArgs.put("account", invoice.getInvoiceNumber());
      newArgs.put("transId", invoice.getPaymentId());
      newArgs.put("note", invoice.getNote());

        String amount = ((Currency) currencyDAO.find(invoice.getDestinationCurrency()))
          .format(invoice.getAmount());

        newArgs.put("currency", invoice.getDestinationCurrency());
        newArgs.put("amount", amount);

        SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MMM-yyyy");
        newArgs.put("date", date != null ? dateFormat.format(date) : "n/a");

        if ( invoice instanceof BillingInvoice ) {
          BillingInvoice billingInvoice = (BillingInvoice) invoice;
          newArgs.put("billingPeriod", String.format("from %s to %s",
            dateFormat.format(billingInvoice.getBillingStartDate()),
            dateFormat.format(billingInvoice.getBillingEndDate()))
          );
        }

        return newArgs;
      `
    },
    {
      name: 'findApproversOftheBusiness',
      type: 'List<UserUserJunction>',
      args: [
        {
          name: 'x',
          type: 'Context'
        }
      ],
      javaCode: `
        // Need to find all approvers and admins in a business
        DAO agentJunctionDAO = (DAO) x.get("agentJunctionDAO");
        User user = ((Subject) x.get("subject")).getUser();
        if ( user == null ) {
          ((Logger) x.get("logger")).error("@InvoiceNotificationDAO and context user is null", new Exception());
          return null;
        }

        // currently the only sme group that can not approve an invoice is employees
        return ((ArraySink) agentJunctionDAO
          .where(
            AND(
              EQ(UserUserJunction.TARGET_ID, user.getId()),
              OR(
                CONTAINS_IC(UserUserJunction.GROUP, "admin"),
                CONTAINS_IC(UserUserJunction.GROUP, "approver")
                )
              )
            ).select(new ArraySink())).getArray();
      `
    }
  ]
});
